terraform/monitoring/03_service_slo.tf

```diff
@@ -0,0 +1,100 @@
+# The SLO is defined as following: 
+#   80% of all non-4XX requests within the past 30 day windowed period
+#   return with 200 OK status
+resource "google_monitoring_slo" "frontend_availability_slo" {
+  service = google_monitoring_custom_service.frontend.service_id
+  slo_id = "availability-slo"
+  display_name = "Availability SLO with request base SLI (good total ratio)"
+
+  # The goal sets our 80% objective with a 30 day rolling window period
+  goal = 0.8
+  rolling_period_days = 30
+
+  request_based_sli {
+    good_total_ratio {
+
+      # The "good" service is the number of 200 OK responses
+      good_service_filter = join(" AND ", [
+        "metric.type=\"istio.io/service/server/request_count\"",
+        "resource.type=\"k8s_container\"",
+        "resource.label.\"cluster_name\"=\"stackdriver-sandbox\"",
+        "metadata.user_labels.\"app\"=\"frontend\"",
+        "metric.label.\"source_workload_namespace\"=\"default\"",
+        "metric.label.\"response_code\"=\"200\""
+      ])
+
+      # The total is the number of non-4XX responses
+      # We elimiate 4XX responses since they do not accurately represent server-side 
+      # failures and have the possibility of skewing our SLO measurements
+      total_service_filter = join(" AND ", [
+        "metric.type=\"istio.io/service/server/request_count\"",
+        "resource.type=\"k8s_container\"",
+        "resource.label.\"cluster_name\"=\"stackdriver-sandbox\"",
+        "metadata.user_labels.\"app\"=\"frontend\"",
+        "metric.label.\"source_workload_namespace\"=\"default\"",
+        "metric.label.\"response_code\"!=\"400\"",
+        "metric.label.\"response_code\"!=\"404\"",
+        "metric.label.\"response_code\"!=\"426\""
+      ])
+      
+    }
+  }
+}
+
+# Here we create another SLO on the frontend service this time with respect to latency
+# Our SLO is defined as following:
+#   80% of requests that return 200 OK responses return in under 500 ms
+resource "google_monitoring_slo" "frontend_latency_slo" {
+  service = google_monitoring_custom_service.frontend.service_id
+  slo_id = "latency-slo"
+  display_name = "Latency SLO with request base SLI (distribution cut)"
+
+  # Again we set our goal of 80% with a 30 day rolling window period.
+  goal = 0.8
+  rolling_period_days = 30
+
+  request_based_sli {
+    distribution_cut {
+
+      # The distribution filter retrieves latencies of reuqests that returned 200 OK responses
+      distribution_filter = join(" AND ", [
+        "metric.type=\"istio.io/service/server/response_latencies\"",
+        "resource.type=\"k8s_container\"",
+        "resource.label.\"cluster_name\"=\"stackdriver-sandbox\"",
+        "metric.label.\"response_code\"=\"200\"",
+        "metadata.user_labels.\"app\"=\"frontend\"",
+        "metric.label.\"source_workload_namespace\"=\"default\""
+      ])
+
+      range {
+
+        # By not setting a min value, it is automatically set to -infinity
+        max = 500
```

Comment:
What's the unit here?  Milliseconds?

Reply:
Yes the units are milliseconds. Added a comment to make this clear to the reader.